### ISSUE TITLE
Crossing the streams: How to manage courses across time

### ISSUE THREAD

**Issue by ritjoe** [FOSSRIT/foss-minor#2]
So, the idea of a *course* is an abstraction. It refers to a number of interlocking things, including:

* a particular instance of a course, which occurs only once ever across a specific period of time at a specific institution, eg, the HFOSS course held during the Spring 2016 semester at RIT
* the more abstract idea of that course as it exists within a curriculum, encompassing instances that have run in the past, one or more ongoing instances, and future instances all taken together, eg, the HFOSS course.
* the specific set of resources used in each course, including the syllabus, course content, and various policies, procedures, hardware, services and so forth, either within a specific instance, or over time.

Over time, instructors and students of the RIT FOSS minor core courses have put a great deal of the material they use into git repositories hosted on GitHub, of which various clones, branches, or forks have been made by both instructors and (as part of course procedure and policy) students. 

Instead of keeping around a repository for each semester, we should, in principle, be able to merge at least the instructors' branches of a course as it develops from term to term back into a single repository.

Further, it seems to me we ought to be able to preserve the ability to refer to a specific terms state at useful, specific time points by applying a git tag to the repository at each of those times.

Now, I've not really got much experience with tagging. Checking one of my most recent clones of the HFOSS repository as I've been using it, forked from https://github.com/decause/HFOSS about a year ago, I see it has no tags at all yet.

From [the reading I've been doing](https://git-scm.com/book/en/v2/Git-Basics-Tagging) I understand that tags can be arbitrary, and that they sort as text. Tagging a course repo with traditional monotonically-increasing version numbers doesn't seem to make much sense to me, at least given their strong association with refinement, added features, bug fixes, etc. Courses don't necessarily work that way. My course is different from those taught by @itprofjacobs, @threebean, @afreshvegetable, or @decause but it's not necessarily a more refined or more featureful version of them.

Instead, I'm considering tagging things using the [ISO 8601](https://en.wikipedia.org/wiki/ISO_8601) date, at least near the start and after the end of each semester. This would provide a very short list of reference points into the code that will sort in an obvious and useful way, that can be compared easily against the academic calendar to place them in an appropriate context.

It might be that we could then incorporate tagging into the courses themselves, possibly to indicate certain assignment due-dates or other high-water marks within each instance of a course.

Does this make any sense at all? Problems? Opportunities?

**Comment by decause** [comment]
Will defer to current Prof's preferences.

So long as we're not worried that the repository will grow to the point where we cannot easily clone it (already in the 100MB+) I'd be interested in exploring this philosophy more.

I think that @ryansb and even moreso, @ralphbean will have interesting opinions here.

**Comment by itprofjacobs** [comment]
It doesn’t seem to be a bad thing to me.  Getting it backwards compatible would be a pain in the ass.  Going forward not a big issue though setting it up so folks know the format and/or that it gets tagged this way automagically would be the win

**Comment by liam-middlebrook** [comment]
I think that tagging based on the academic term 2131 might make more sense if we want to know the course content and state for a particular instance.

**Comment by liam-middlebrook** [comment]
2131 being an example of a term id. Possibly we would say term-$termid

**Comment by ritjoe** [comment]
On Mon, Aug 08, 2016 at 12:40:09PM -0700, Liam Middlebrook wrote:
> 2131 being an example of a term id. Possibly we would say term-$termid

I have very few data points (two from last year and the
upcoming term for a grand total of 3!) to inform my experience about this.

Would we then append a third token to indicate state throughout a term, eg,
-begin, -end? Those two would sort ok I guess, but I don't know about other
components in a tag we might want to use throughout the term.

Do you think it makes any sense to try to tag things within a term?

-- 
D. Joe Anderson                        deejoe@mail.rit.edu
Adjunct Faculty              (585) 678-1563 (google voice)   
Interactive Games & Media                       GCCIS, RIT

**Comment by ritjoe** [comment]
On Mon, Aug 08, 2016 at 01:32:58PM -0700, Joe wrote:
> On Mon, Aug 08, 2016 at 12:40:09PM -0700, Liam Middlebrook wrote:
> > 2131 being an example of a term id. Possibly we would say term-$termid

> Would we then append a third token to indicate state throughout a term, eg,
> -begin, -end? Those two would sort ok I guess, but I don't know about other
> components in a tag we might want to use throughout the term.

Another consideration is transparency and ease of external reference. What I
mean by this is, how might it be used in practice?  If tags were date
strings, and one had a date tag, one might be able to guess which semester
but not know if it was before, during, or after. And of course, this would
be harder for the courses done under the quarter system, if we ever got
around to tagging them.

So, to try to find out where something falls relative to a recent year, do a
search for "RIT calendar" or "RIT academic calendar".  Top hits include
links to the current or upcoming calendar and the current or previous
calendar (eg, this years and last years, or this years and next years,
depending).

Turns out this is not so good, since there's no obvious general mechanism on

https://www.rit.edu/calendar/

for pulling up an arbitrary calendar.

In contrast, taking the same approach for course numbers or term IDs, I did
the following.

Calling up "RIT course descriptions" the top hit I find is a link to a
current PDF, which is OK.  

But what is really cool here is the second link:

[Office of the Registrar: Course Catalogs](https://www.rit.edu/academicaffairs/registrar/course-catalogs)

This is a link into a digital archive giving a list of years going back to
the 1970s, and within that year a short list of documents that can be
retrieved by PDF. Selecting a year,document pair then leads to a landing page
from which a PDF can be opened or downloaded. (As an aside, what's particularly relevant
to our interests here otherwise is that I've seen at least two that have
some sort of statement on the landing page that the documents are available
under a creative commons license).

I didn't find this in the year before, but did find this as what I think is
the first catalog mention of "open source":

https://digitalarchive.rit.edu/xmlui/bitstream/handle/1850/14170/Undergrad_Course_Descriptions.pdf?sequence=1

Amongst all the other information, these documents have calendars, term ids,
and course numbers.

I'm not sure this settles anything, but it includes a few nuggets of shiny
relevance I hadn't anticipated finding.
